modules/Microsoft.Compute/galleries/images/deploy.bicep

```diff
@@ -133,7 +141,13 @@ resource image 'Microsoft.Compute/galleries/images@2021-10-01' = {
         max: maxRecommendedMemory
       }
     }
-    hyperVGeneration: hyperVGeneration
+    hyperVGeneration: !empty(securityType) ? 'V2' : hyperVGeneration
```

Comment:
I would definitely be interested in your comments about this type of thing where we have parameter dependencies in combination with allowed values and default values, etc. This module isn't such a big deal as it is just an image definition and I feel as though if someone specifically chose to call out a securityType that is what they want so changing the hyperVGeneration would be acceptable.

However, I see where we have bigger concerns or conversations when talking about the same types of dependencies on VirtualMachines or other resources. I've specifically struggled with this with the CARML repo and the AVD Accelerator code. Where do we put logic to prevent user error? Does the logic go in the Module or does it go in a deployment BICEP file or other pipeline artifact that dynamically changes the parameters when this type of dependency error can occur? In my humble opinion, the key is to try to maintain consistency and document the approach so anyone using the CARML understands that it will/will not correct these types of errors.

Reply:
From my perspective, if I have two parameter files with different configurations, I would prefer to have 'hyperVGeneration' explicitly called out as 'V2' when the user wants to use 'securityType' opposed to abstracting the logic in the module. It's easier for someone to understand the configuration by just looking at the parameter file. So I'd go with "making it fail" if not configured properly

Reply:
It's a tough nut to crack. Personally, in this case I'd prefer the implemented logic, as the security type seems to enforce V2. So if somebody runs the module with the default settings (that currently specificy V1), it would work like a charm.

However, I can't say I have a strong opinion about it either. It's unfortunately not a parameter where we could simply say 'if a user specified something we'll go with then, otherwise we try to find the best value' - like we do with the `publicAccess` param in several modules.

Reply:
@AlexanderSehr actually we could, if we added empty (`''`) to the `hyperVGeneration` allowed values, set that as default value, and build here a condition such as 
```suggestion
    hyperVGeneration: !empty(hyperVGeneration) ? hyperVGeneration : (!empty(securityType) ? 'V2' : 'V1')
```
This would allow users to enforce values if explicitly set through parameters, otherwise to rely on what we try to set as the "good default"

Reply:
I am not sure allowing it to be blank is the right option here. When we create a new image definition from the portal it defaults to V1 and switches to V2 automatically if you select one of the advanced securityTypes. I know there isn't a way to do that in a template exactly, but if we document the template behavior, shouldn't we provide the same functionality of the portal. I took the same approach with the VirtualMachine module update to automatically set the System-Assigned identity to enabled if the user enabled the AAD login extension.

Reply:
I get where you're coming from, but actually no :D When talking about best pratices / secure parameters we got the feedback (and came to the conclusion) that we should not consider the portal values as deliberatly 'good' defaults. If there is nothing else to go on, I'd probably also refer to those, but they are unfortunately not meaningful enough for anything else.

Regarding the empty value @eriqua, that's actually a good idea. I agree with @shawntmeyer that this would naturally diverge from the API - but - it's not a unique thing in CARML. We actually use similar techniques in several places to ease the usage of the module. Be it to have `-1` as a default for an integer (to then consider it 'not provided', or in this case an empty string.

The idea in this case would anyways be to make it easier to deploy, the same was as we introduce e.g. custom, CARML-specific CMK parameters that are aligned across our modules to make it easier to deploy (as RPs tend to all implement their own interpretation).